File: scripts/controller_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from nav_msgs.msg import Odometry  # Required for reading position
from assignment2_rt.srv import GetVelocity  # Import from YOUR package
import threading


class ControllerNode(Node):
    def __init__(self):
        super().__init__('controller_node')

        # Subscribe to /odom to track position
        self.create_subscription(Odometry, '/odom', self.odom_callback, 10)

        self.publisher_ = self.create_publisher(Twist, '/cmd_vel', 10)
        self.srv = self.create_service(GetVelocity, 'get_avg_velocity', self.get_avg_callback)

        self.current_twist = Twist()
        self.vel_history = []
        self.robot_x = 0.0  # Track X position

        # --- LIMITS ---
        self.LIMIT_X_MAX = 3.0
        self.LIMIT_X_MIN = -3.0

        self.timer = self.create_timer(0.1, self.timer_callback)
        self.get_logger().info(
            f"Controller Started. Constraints: {
                self.LIMIT_X_MIN} < X < {
                self.LIMIT_X_MAX}")

    def odom_callback(self, msg):
        """Updates the robot's current X position."""
        self.robot_x = msg.pose.pose.position.x

    def timer_callback(self):
        """Checks limits and publishes velocity."""
        safe_cmd = Twist()
        safe_cmd.linear.x = self.current_twist.linear.x
        safe_cmd.angular.z = self.current_twist.angular.z

        # --- LIMIT LOGIC ---
        # 1. If at MAX limit (3.0) AND trying to go forward -> STOP
        if self.robot_x >= self.LIMIT_X_MAX and safe_cmd.linear.x > 0:
            self.get_logger().warn(
                f"LIMIT REACHED (X={
                    self.robot_x:.2f})! Stopping forward.",
                throttle_duration_sec=2)
            safe_cmd.linear.x = 0.0

        # 2. If at MIN limit (-3.0) AND trying to go backward -> STOP
        elif self.robot_x <= self.LIMIT_X_MIN and safe_cmd.linear.x < 0:
            self.get_logger().warn(
                f"LIMIT REACHED (X={
                    self.robot_x:.2f})! Stopping backward.",
                throttle_duration_sec=2)
            safe_cmd.linear.x = 0.0

        self.publisher_.publish(safe_cmd)

    def set_velocity(self, lin_x, ang_z):
        self.current_twist.linear.x = float(lin_x)
        self.current_twist.angular.z = float(ang_z)

        self.vel_history.append((lin_x, ang_z))
        if len(self.vel_history) > 5:
            self.vel_history.pop(0)

    def get_avg_callback(self, request, response):
        if not self.vel_history:
            response.avg_linear = 0.0
            response.avg_angular = 0.0
            return response

        sum_lin = sum(v[0] for v in self.vel_history)
        sum_ang = sum(v[1] for v in self.vel_history)

        response.avg_linear = sum_lin / len(self.vel_history)
        response.avg_angular = sum_ang / len(self.vel_history)
        return response


def main(args=None):
    rclpy.init(args=args)
    node = ControllerNode()
    spin_thread = threading.Thread(target=rclpy.spin, args=(node,), daemon=True)
    spin_thread.start()

    try:
        while rclpy.ok():
            print("\n" + "=" * 40)
            print(f" CONTROLLER INTERFACE")
            print(f" Status: X = {node.robot_x:.2f}")
            print(f" Limits: [{node.LIMIT_X_MIN}, {node.LIMIT_X_MAX}]")
            print("=" * 40)
            l_input = input("Enter Linear Velocity (x): ")
            a_input = input("Enter Angular Velocity (z): ")
            try:
                l_vel = float(l_input)
                a_vel = float(a_input)
                node.set_velocity(l_vel, a_vel)
                print(f"Target Set: v={l_vel}, w={a_vel}")
            except ValueError:
                print("Invalid Input.")
    except KeyboardInterrupt:
        print("\nExiting...")
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/controller_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "assignment2_rt/srv/get_velocity.hpp"

#include "assignment2_rt/srv/set_threshold.hpp"

#include <iostream>
#include <vector>
#include <numeric>
#include <thread>
#include <deque>
#include <limits>
#include <mutex>

// Using std::deque for the history buffer (FIFO)
struct VelCommand {
    float linear;
    float angular;
};

class ControllerNode : public rclcpp::Node
{
public:
    ControllerNode() : Node("controller_node")
    {
        // 1. Publisher
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);

        // 2. Service Server
        service_ = this->create_service<assignment2_rt::srv::GetVelocity>(
            "get_avg_velocity",
            std::bind(&ControllerNode::get_avg_callback, this, std::placeholders::_1, std::placeholders::_2)
        );

        // 3. Service Client
        client_threshold_ = this->create_client<assignment2_rt::srv::SetThreshold>("set_safety_threshold");

        RCLCPP_INFO(this->get_logger(), "Controller Node (C++) Started. Ready for input.");
    }

    void publish_velocity(float lin_x, float ang_z)
    {
        auto msg = geometry_msgs::msg::Twist();
        msg.linear.x = lin_x;
        msg.angular.z = ang_z;
        publisher_->publish(msg);

        update_history(lin_x, ang_z);
    }

    void call_threshold_service(float new_thresh)
    {
        if (!client_threshold_->wait_for_service(std::chrono::seconds(1))) {
            RCLCPP_WARN(this->get_logger(), "Service 'set_safety_threshold' not available.");
            return;
        }

        auto request = std::make_shared<assignment2_rt::srv::SetThreshold::Request>();
        request->new_threshold = new_thresh;

        // Call asynchronously and ignore result for now (simplification)
        client_threshold_->async_send_request(request);
    }

private:
    void update_history(float lin, float ang)
    {
        std::lock_guard<std::mutex> lock(history_mutex_);
        vel_history_.push_back({lin, ang});
        if (vel_history_.size() > 5) {
            vel_history_.pop_front();
        }
    }

    void get_avg_callback(const std::shared_ptr<assignment2_rt::srv::GetVelocity::Request> request,
                          std::shared_ptr<assignment2_rt::srv::GetVelocity::Response> response)
    {
        (void)request; // Avoid unused parameter warning

        std::lock_guard<std::mutex> lock(history_mutex_);
        if (vel_history_.empty()) {
            response->avg_linear = 0.0;
            response->avg_angular = 0.0;
            RCLCPP_WARN(this->get_logger(), "Request received, but history is empty.");
            return;
        }

        float sum_lin = 0.0;
        float sum_ang = 0.0;

        for (const auto& cmd : vel_history_) {
            sum_lin += cmd.linear;
            sum_ang += cmd.angular;
        }

        response->avg_linear = sum_lin / vel_history_.size();
        response->avg_angular = sum_ang / vel_history_.size();

        RCLCPP_INFO(this->get_logger(), "Service called. Avg Lin: %.2f, Avg Ang: %.2f",
                    response->avg_linear, response->avg_angular);
    }

    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::Service<assignment2_rt::srv::GetVelocity>::SharedPtr service_;
    rclcpp::Client<assignment2_rt::srv::SetThreshold>::SharedPtr client_threshold_;
    std::deque<VelCommand> vel_history_;
    std::mutex history_mutex_;
};

// --- INPUT LOOP FUNCTION ---
void user_input_loop(std::shared_ptr<ControllerNode> node)
{
    float l_vel, a_vel, thresh;
    while (rclcpp::ok()) {
        std::cout << "\n---------------------------------" << std::endl;
        std::cout << "--- Robot Controller Interface (C++) ---" << std::endl;

        std::cout << "Enter Linear Velocity (x): ";
        if (!(std::cin >> l_vel)) {
            std::cin.clear(); std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            continue; // Handle bad input
        }

        std::cout << "Enter Angular Velocity (z): ";
        if (!(std::cin >> a_vel)) {
            std::cin.clear(); std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            continue;
        }

        std::cout << "Enter Safety Threshold (m): ";
        if (!(std::cin >> thresh)) {
            std::cin.clear(); std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            continue;
        }

        node->publish_velocity(l_vel, a_vel);
        node->call_threshold_service(thresh);

        std::cout << "Sent: Linear=" << l_vel << ", Angular=" << a_vel << ", Threshold=" << thresh << std::endl;
    }
}

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ControllerNode>();

    // Run the user input loop in a separate thread so ROS callbacks (service) can run in main
    std::thread input_thread(user_input_loop, node);

    rclcpp::spin(node);

    // Cleanup
    if (input_thread.joinable()) input_thread.detach();
    rclcpp::shutdown();
    return 0;
}
